tls: throw an error on getLegacyCipher

jasnell opened the pull request:
throw an error on unknown getLegacyCipher input
rather than returning the default. Makes the
behavior more explicit and matches up with the
behavior on the command line switches.

jasnell committed:
disable RC4, add --cipher-list command line switch

Disable RC4 in the default cipher list

Add the `--cipher-list` command line switch and `NODE_CIPHER_LIST`
environment variable to completely override the default cipher list.

Add the `--enable-legacy-cipher-list` and `NODE_LEGACY_CIPHER_LIST`
environment variable to selectively enable the default cipher list from
previous node.js releases.

Reviewed-By: James M Snell <jasnell@gmail.com>
PR-URL: https://github.com/joyent/node/pull/14414

jasnell commented:
@misterdjules @mdawsonibm

misterdjules commented:
Can we add a test to make sure that getLecagyCiphers throws/doesn't throw as expected?

jasnell commented:
Done

jasnell commented:
https://github.com/joyent/node/issues/15445 @misterdjules

mdawsonibm commented:
One minor nit.  Should the tests validate the type of the Error, ie a TypeError or an Error as appropriate ? Otherwise lgtm.

jasnell commented:
/cc @mhdawson @misterdjules ok, please check the two additional commits. if these look good I'll squash the pr down and land.

jasnell commented:
@misterdjules @mhdawson ... ok, squashed the commits and made the additional changes discussed

misterdjules commented on doc/api/tls.markdown:
```diff
@@ -215,10 +244,16 @@ Example:
+release version. The only values currently supported are `v0.10.38`,
+`v0.10.39`, and `v0.12.2`.
 
-Example:
```
Keeping an example (with code formatting) would be nice.

misterdjules commented on test/simple/test-tls-cipher-list.js:
```diff
@@ -43,6 +43,54 @@ function doTest(checklist, env, useswitch) {
       // Test NODE_LEGACY_CIPHER_LIST
       if (env) options = {env:{"NODE_LEGACY_CIPHER_LIST": env}};
       break;
+    case 4:
```
Having arbitrary numbers determine the behavior of tests seems very confusing to me, because it's not clear when I read `doTest(V1038Ciphers, 'v0.10.38', 8);` that we test that `NODE_LEGACY_CIPHER_LIST takes precendence over NODE_CIPHER_LIST`.

Can we refactor that to make it clearer? Maybe we could have separate function clearly named for each test we want to perform? Parameters for each test would be local to these functions and make it clear what each test does.

misterdjules commented on lib/_tls_wrap.js:
```diff
@@ -834,10 +834,13 @@ exports.connect = function(/* [port, host], options, cb */) {
     checkServerIdentity: tls.checkServerIdentity
   };
 
+  if (tls.DEFAULT_CIPHERS != tls.getLegacyCiphers('v0.10.38')) {
```
If one passes `--cipher-list='ECDHE-RSA-AES128-SHA256:AES128-GCM-SHA256:RC4:HIGH:!MD5:!aNULL:!EDH'`, my understanding is that `tls.DEFAULT_CIPHERS` will === `tls.getLegacyCiphers('v0.10.38'))`, and `default.ciphers` won't be set. I haven't tested it, so it may be a wrong assumption.

So I guess that here what is meant is more `if (user didn't pass --enable-legacy-cipher-list=v0.10.38) { ... }`, which is not equivalent to `if (tls.DEFAULT_CIPHERS != tls.getLegacyCiphers('v0.10.38'))`.

Thoughts?

misterdjules replied:
Also, in general please use strict equality unless not possible. In this occurence, please use `!==` instead of `!=`.

jasnell committed:
tls: cipher-list command line switches, test case

Add the `--cipher-list` and `--enable-legacy-cipher-list` command
line switches.

Add the `NODE_CIPHER_LIST` and `NODE_LEGACY_CIPHER_LIST` environment
variables.

Add the getLegacyCiphers method to tls.js

Add the test/simple/test-tls-cipher-list.js test case

(this is a reworked and squashed recommit of the work previously
backed back out of v0.12 for the v0.12.3 release)

jasnell commented:
@misterdjules @mhdawson ... ok, the PR has been rebased following the v0.12.3 cut and squashed down to just two commits against the v0.12 branch. @misterdjules please review and let me know if you feel this is ready to land.

jasnell commented:
Closing this. Taking a different approach.
